Add a time synchronization page (#514)
xcp-ng/xcp-ng-org: docs/management/time-synchronization.md

---
sidebar_position: 8
---

> # Time synchronization

Reviewer: Unlike French (which loves nouns and noun phrases) the English language loves action and verbs. They sound more natural. As a result, I suggest something like **Synchronizing time**, or **Keeping time in sync**.

Author: Your point about English preferring verbs is fair, and I nearly took it. What stopped me is that the neighbours are all noun phrases: Additional packages, Backup, Cloud features, High availability, Monitoring and alerts, Updates, Users and permissions. Renaming this one makes it the odd entry in its own sidebar.

If the section should move to verb phrases, I would rather do that as its own change across all of them, and I would support it.

Reviewer: True; we should keep the wording consistent with the neighbor pages indeed.

> XCP-ng keeps the dom0 clock with `chronyd`. A host whose date is wrong carries on working

Reviewer: ```suggestion
XCP-ng keeps the dom0 clock synchronized using `chronyd`. A host whose date is wrong keeps working
```

Author: Applied. "Synchronized using" is more accurate than "keeps the dom0 clock with", which was vague about what chronyd actually does.

> normally until something needs to trust that date, and the failures it produces then rarely

Reviewer: ```suggestion
normally until something needs to trust that date. When that happens, the resulting failures rarely
```

Author: Applied. Splitting it at "When that happens" fixes the run-on I had.

> mention time at all. This page covers why the clock matters, how to configure NTP, and how

Reviewer: ```suggestion
point to the clock as the cause. This page explains why the clock matters, how to configure NTP, and how
```

Author: Applied. "Point to the clock as the cause" is the better phrase: it says which cause is being missed, which was the whole point of the sentence.

to check and correct it.

## ⏰ Why the clock must be correct {#why-the-clock-must-be-correct}

- **Access to the repositories.** Updates are fetched over HTTPS, and TLS validates the

>   server's certificate against the host's own date. A host set far in the past cannot

Reviewer: ```suggestion
If the host's clock is set too far in the past, the TLS handshake fails because, from the host's perspective, the mirror's certificate is not yet valid.
```

Author: Applied, with the last sentence kept: "The error `yum` prints in that situation does not mention time." It is the reason the troubleshooting entry in #504 exists, so I did not want to lose it.

  complete the handshake, because from where it is standing the mirror's certificate is not
  valid yet. The error `yum` prints in that situation does not mention time.
- **Pool membership.** A host joining a pool must have its clock synchronized with the pool
  master. See [Pool Requirements](../../installation/requirements#pool-requirements).
- **Certificates.** Certificates are only valid inside a validity window. A host whose date
  falls outside that window will reject, or be rejected by, connections that would otherwise
  succeed.
- **Logs and scheduled operations.** Correlating events across the hosts of a pool, and
  anything that runs on a schedule such as backups, depend on the hosts agreeing on what
  time it is.

## 🕰️ How XCP-ng keeps time {#how-xcp-ng-keeps-time}

The time daemon on dom0 is `chronyd`, configured through `/etc/chrony.conf`.

> The shipped configuration enables `rtcsync`, so chronyd keeps the hardware clock in step

Reviewer: ```suggestion
The default configuration enables rtcsync, allowing chronyd to keep the hardware clock synchronized with the system clock automatically. There is no need to manually update the hardware clock after correcting the system time, and the correction persists across reboots.
```

Author: Applied. You suggested almost exactly this on #504 as well, and that is the copy I owed you: the NTP material was duplicated across both PRs, and it now lives only here.

with the system clock on its own. There is no need to copy a corrected time to the hardware
clock by hand, and a correction survives a reboot.

## 🔧 Configuring the time sources {#configuring-the-time-sources}

### During installation

> The installer asks for the timezone and the time at step 11, and lets you either give it NTP

Reviewer: Having 'and' twice in quick succession in the same sentence sounds awkward. I suggest the following:

```suggestion
At step 11, the installer prompts you to enter the time, select a time zone and either configure NTP servers or set the time manually. Always configure at least one NTP server.
```

Author: Applied. Two "and"s in one sentence was clumsy, and your version also puts the instruction last where it belongs.

servers or set the time manually. Always give it an NTP server. See
[Install XCP-ng](../../installation/install-xcp-ng).

:::warning
If you set the time manually at this step, the host can end up with `chronyd` running but no
time source configured at all. It will keep whatever date it was installed with. See
[Checking the time sources](#checking-the-time-sources) below.
:::

### During an automated installation

The installation answer file accepts one or more NTP servers through the `<ntp-server>`
element. See [Answer file](../../appendix/answerfile).

### On an installed host

You can configure NTP from `xsconsole`, or by editing `/etc/chrony.conf` directly and
restarting the service.

## 🔍 Checking the time sources {#checking-the-time-sources}

Check that the service is running, and that it actually has something to synchronize with:

```bash
systemctl status chronyd
chronyc sources
```

Both matter, and the second is the one that gets skipped. `chronyd` can be running and
enabled while having **no time sources configured at all**:

```
# chronyc sources
210 Number of sources = 0
```

> The service is then behaving exactly as configured, and doing nothing. Every other check an

Reviewer: ```suggestion
Then, the service behaves exactly as configured, and does nothing. Every other check an
```

Author: Applied.

operator would normally run looks healthy, which is what makes this one worth running early
rather than late.

When sources are configured, `chronyc sources` lists them. In the `MS` column, the second

> character is the state of that source: `*` marks the one currently being used, `+` marks

Reviewer: ```suggestion
character is the state of that source: 
 - `*` marks the one currently being used. 
 - `+` marks another acceptable source being combined with it.
 - `?` means the source is unreachable. 
```

Author: Applied as a list. Three states buried in a running sentence was exactly the wrong shape for something a reader scans while looking at their own output.

another acceptable source being combined with it, and `?` means the source is unreachable.
At least one source should reach the `*` state.

To see the offset chrony believes it has, and whether it has settled:

```bash
chronyc tracking
```

## 🛠️ Correcting a wrong clock {#correcting-a-wrong-clock}

Start with `chronyc sources`, because the answer decides which of the two paths below
applies.

### If sources are listed

Correct the clock and check the result:

```bash
chronyc makestep
date
```

> `chronyc makestep` is what does the work here. By default chrony corrects an offset by

Reviewer: ```suggestion
`chronyc makestep` is what does the work here. By default, chrony corrects time offsets by gradually slewing the system clock, which never converges for an offset of months or years.
```

Author: Applied. This is the same sentence you flagged on #504, where you also pointed out that "It" had lost its referent by the time the reader arrived. Both fixed here, in the copy that survives.

slewing the clock gradually, which never converges for an offset of months or years.

### If `Number of sources = 0`

Do not start with `makestep`. It will report success and move nothing, because chrony has no

> measured offset to step to. Add time sources to `/etc/chrony.conf` first:

Reviewer: ```suggestion
measured offset to step to. 

1. Add time sources to `/etc/chrony.conf`:
```

Author: Applied, as a numbered procedure.

One departure worth flagging: I indented the code blocks under their list items. Left at column zero, each fence closes the list and the following number starts a new one, so it renders as three lists rather than three steps. Same output you asked for, just held together.

```

> server 0.centos.pool.ntp.org iburst

Reviewer 2: ```suggestion
server 2.pool.ntp.org iburst
```

I think it's preferable to not use centos vendor zone. since XCP-ng is a different vendor.

I think it's acceptable for users to use a public pool, as suggested at:

https://community.ntppool.org/t/oracle-linux-vendor-zone-request-for-a-clear-decision-and-a-usable-vendor-zone-process/4503/5?u=rzr

FYI xcpng vendor zone is still pending (for months):

https://github.com/xcp-ng-rpms/chrony/pull/1#discussion_r3549768407

Author: Taken, and thanks for the links. The vendor zone was not ours to use and I had not thought about it at all. Switched to the generic pool.

Worth someone chasing xcp-ng-rpms/chrony#1 so there is an XCP-ng zone to point at eventually, but this page should not wait on that.

server 1.centos.pool.ntp.org iburst
server 2.centos.pool.ntp.org iburst
server 3.centos.pool.ntp.org iburst
```

> Then restart the service and confirm that a source has become reachable, as described in

Reviewer: ```suggestion
2. Restart the service and confirm that a source has become reachable, as described in
```

Author: Applied, with the continuation line indented under the item so the link stays part of step 2.

[Checking the time sources](#checking-the-time-sources):

```bash
systemctl restart chronyd
chronyc sources
```

> Once a source is reachable, correct the clock:

Reviewer: ```suggestion
3. Once a source is reachable, correct the clock:
```

Author: Applied.

```bash
chronyc makestep
date
```

:::note
Correcting the date does not invalidate the host's own certificate. XAPI issues it with a
ten-year validity, so a host installed with a wrong date still holds a certificate that
covers the corrected date. `xe host-refresh-server-certificate` is not needed for this.
:::

:::tip
If `date` is wrong again after every power cycle, the motherboard's RTC battery is probably
dead and should be replaced. `rtcsync` can only keep the hardware clock in step while the
host is running; it cannot help a clock that loses its value when the power goes.
:::

> ## 🌐 Isolated networks {#isolated-networks}

Reviewer: 'Isolated' sounds a bit French to me (due to its latin origin). Would 'Offline' work in this case?

Author: I would keep "isolated" here, for a reason beyond taste: "offline" already means something else in these docs. It is used for a host or a server being down, in `ha.md`, `xostor.md` and `migrate-to-xcp-ng.md`. Meanwhile "isolated network" is already the phrasing in `management/backup.md`.

Reusing "offline" for "no internet access" would collide with the sense readers have already met elsewhere.

Reviewer 2: what about offgrid ?

Author: "Offgrid" I would avoid for the same reason as "offline": it appears nowhere else in these docs, and it reads informally for a page someone lands on while their host is broken.

But two of you tripping on the same word is a signal even when the replacements do not work, so I have stopped arguing for "isolated" and named the condition instead: **Networks without internet access**. It says exactly who the section is for and borrows nobody else's meaning. Nothing linked to the old anchor, so nothing breaks.

Reviewer: **Networks without internet access** is longer and therefore a little less elegant, but it also avoids any ambiguity. Seems fine for a technical documentation. I approve!

> Hosts without access to the internet cannot reach the public NTP pool, and the symptom is

Reviewer: ```suggestion
Hosts without Internet access cannot reach the public NTP pool. The result is the same as having no sources at all:
```

Author: Applied. Splitting it into two sentences reads better than the comma splice I had.

the same as having no sources at all: `chronyc sources` lists servers that never leave the
`?` state.

> On such a network, point `/etc/chrony.conf` at a time source that hosts can actually reach,

Reviewer: ```suggestion
On such networks, point `/etc/chrony.conf` at a time source that hosts can actually reach,
```

Author: Applied.

such as an appliance on the same network or a local server that is itself synchronized.

> ## 🎱 Pools {#pools}

Reviewer: Are we sure that an 8-ball is the best emoji to represent VM pools? Is the metaphor going to be understood clearly for all readers, from different parts of the world and cultural backgrounds?

Author: The metaphor concern is reasonable and I do not think you are wrong about it. The snag is that the 8-ball is already this site's pool emoji: `guides/guest-UEFI-Secure-Boot.md` and `installation/requirements.md` both use it for pool sections. Changing it here alone makes this page the inconsistent one.

Happy to follow if you want to raise it site-wide.

Reviewer: Thanks for the observation. Let's keep the emoji as-is.

Keep every host in a pool synchronized, ideally against the same sources. A host whose clock

> disagrees with the pool master is not merely inconvenient: clock synchronization is one of

Reviewer: ```suggestion
differs from the pool master is not just inconvenient: clock synchronization is one of
```

Author: Applied. "Differs from" is the plainer verb.

> the requirements for joining a pool in the first place, alongside the others listed in

Reviewer: I would make the sentence a little lighter, like this:

```suggestion
the requirements for joining a pool, along with the other requirements listed in
```

Author: Applied. Lighter, and it drops the "in the first place" I had leaned on.

---
sidebar_position: 8
---

# Time synchronization

XCP-ng keeps the dom0 clock synchronized using `chronyd`. A host whose date is wrong keeps
working normally until something needs to trust that date. When that happens, the resulting
failures rarely point to the clock as the cause. This page explains why the clock matters, how
to configure NTP, and how to check and correct it.

## ⏰ Why the clock must be correct {#why-the-clock-must-be-correct}

> - **Access to the repositories.** Updates are fetched over HTTPS, and TLS validates the

Reviewer 2: I would put this item at end of this list, since a non accurate time is not too annoying for https

Author: This is the one I would push back on. An inaccurate clock does not merely annoy HTTPS: if the host's date falls outside the certificate's validity window, the handshake fails outright and nothing is fetched. I tested that on a host earlier today, and it is the whole subject of the companion PR #504.

It is also the reason people arrive at this page. Forum threads 11558 and 9560 are both someone whose `yum` stopped working with an error that never mentions time, and the second one ends with nobody working it out. The item is first because it is the failure that hides its own cause, not because it is the most severe.

The pool and certificate items are arguably worse when they bite. They are just easier to recognise.

Reviewer 3: I would still insist on moving it further down. Failure to join update repositories doesn't have a direct impact production, whereas failure to communicate correctly between hosts can prevent you from performing important operations.

operator would normally run looks healthy, which is what makes this one worth running early
rather than late.

When sources are configured, `chronyc sources` lists them. In the `MS` column, the second
character is the state of that source:

- `*` marks the one currently being used.
- `+` marks another acceptable source being combined with it.
- `?` means the source is unreachable.

At least one source should reach the `*` state.

To see the offset chrony believes it has, and whether it has settled:

```bash
chronyc tracking
```

## 🛠️ Correcting a wrong clock {#correcting-a-wrong-clock}

Start with `chronyc sources`, because the answer decides which of the two paths below
applies.

### If sources are listed

Correct the clock and check the result:

```bash
chronyc makestep
date
```

`chronyc makestep` is what does the work here. By default, chrony corrects time offsets by
gradually slewing the system clock, which never converges for an offset of months or years.

### If `Number of sources = 0`

Do not start with `makestep`. It will report success and move nothing, because chrony has no
measured offset to step to.

1. Add time sources to `/etc/chrony.conf`:

   ```

>    server 0.pool.ntp.org iburst

Reviewer 2: ok, let me see if we can have a `server ntp.xcp-ng.org` (doc can be updated afterwards anyway)

Note to self, think to update the doc updated once https://github.com/xcp-ng-rpms/chrony/pull/1 is merged

Author: Good, that would be better than sending people at the public pool. Happy to update the page once xcp-ng-rpms/chrony#1 lands and there is something real to point at.

   server 1.pool.ntp.org iburst
   server 2.pool.ntp.org iburst
   server 3.pool.ntp.org iburst
   ```

2. Restart the service and confirm that a source has become reachable, as described in
   [Checking the time sources](#checking-the-time-sources):

   ```bash
   systemctl restart chronyd
   chronyc sources
   ```

3. Once a source is reachable, correct the clock:

   ```bash
   chronyc makestep
   date
   ```

:::note
Correcting the date does not invalidate the host's own certificate. XAPI issues it with a
ten-year validity, so a host installed with a wrong date still holds a certificate that
covers the corrected date. `xe host-refresh-server-certificate` is not needed for this.
:::

:::tip
If `date` is wrong again after every power cycle, the motherboard's RTC battery is probably
dead and should be replaced. `rtcsync` can only keep the hardware clock in step while the
host is running; it cannot help a clock that loses its value when the power goes.
:::

## 🌐 Isolated networks {#isolated-networks}

Hosts without Internet access cannot reach the public NTP pool. The result is the same as
having no sources at all: `chronyc sources` lists servers that never leave the `?` state.

> On such networks, point `/etc/chrony.conf` at a time source that hosts can actually reach,

Reviewer 2: I would use "point to" instead

Author: Applied.

> such as an appliance on the same network or a local server that is itself synchronized.

Reviewer 2: or will be used as reference time for the network

Author: Taken, that is the more useful half of the sentence. It now reads "a local server that is itself synchronized and acts as the reference time for the network" — being synchronized is not the point on its own, being what the others follow is.

## 🎱 Pools {#pools}

> Keep every host in a pool synchronized, ideally against the same sources. A host whose clock

Reviewer 2: ```suggestion
Keep every host in a pool synchronized, ideally against the same source. A host whose clock
```

Author: I would keep the plural here. Chrony wants several sources so it can tell a bad one from a good one, and the point of the sentence is that the hosts of a pool should agree on the same *set* rather than each picking its own. Singular would quietly recommend a single upstream, which is the weaker setup.

Happy to be argued out of it if you meant something narrower.

Reviewer 2: well afaik there should be only one reference clock in the end, but i understand now you were talking about config file entries, talking about this unsure there is a need to list severals but who knows, better safe than sorry.

Author: Four is what XCP-ng ships. I checked two 8.3 hosts, stock `/etc/chrony.conf` from `chrony-3.2-2.el7`:

```
server 0.centos.pool.ntp.org iburst
server 1.centos.pool.ntp.org iburst
server 2.centos.pool.ntp.org iburst
server 3.centos.pool.ntp.org iburst
```

The page keeps the count and drops the vendor zone, per your earlier comment. So someone rebuilding a config that lost its sources lands on the shape a fresh install would have given them, without the zone that is not ours to use.

On the single reference clock I think we are saying the same thing from two ends. One of those hosts right now:

```
^+ time.cloudflare.com        3  10   377   679   -406us[ -365us] +/-   10ms
^+ ip217-154-21-219.pbiaas.>  2  10   377   701  +2094us[+2135us] +/-   14ms
^* mail.rapidooo.fr           2  10   377   653   -772us[ -731us] +/-   10ms
^- 172-234-184-36.ip.linode>  2   9   377   165  +1864us[+1864us] +/-   31ms
```

One `*`, and the rest are there so chrony can work out whether that one is still worth trusting. Four entries, one reference clock.

That output also has a `-`, and the page only explains `*`, `+` and `?`. Both hosts show one, so a reader following the page will meet a character it does not cover. Add it here, or leave this PR as it is and do it separately?

Author: Correction on my last message: what I pasted was not stock, it was my own repair read back. I edited both those hosts on 28 July when their clocks were wrong.

The package does ship the four lines. Taking an untouched pre-edit copy and putting the four `server` entries back under the header comments reproduces the sha256 that `rpm -q --dump chrony` records for `/etc/chrony.conf`, byte for byte.

Neither installed host had them, though. Both carried a 964 byte file with no `server` or `pool` line at all, byte-identical to each other, installed two months apart. That is `Number of sources = 0` on a host nobody had touched yet, which is the case this page spends a section on. What removes them I do not know, and I would rather not name a cause I have not pinned down.

So four is the package's number rather than mine, which was the thing I was trying to establish. Saying you would find it on an installed host was wrong.

hardware clock after correcting the system time, and the correction persists across reboots.

## 🔧 Configuring the time sources {#configuring-the-time-sources}

### During installation

At step 11, the installer prompts you to enter the time, select a time zone and either
configure NTP servers or set the time manually. Always configure at least one NTP server. See
[Install XCP-ng](../../installation/install-xcp-ng).

:::warning
If you set the time manually at this step, the host can end up with `chronyd` running but no
time source configured at all. It will keep whatever date it was installed with. See
[Checking the time sources](#checking-the-time-sources) below.
:::

### During an automated installation

The installation answer file accepts one or more NTP servers through the `<ntp-server>`
element. See [Answer file](../../appendix/answerfile).

### On an installed host

You can configure NTP from `xsconsole`, or by editing `/etc/chrony.conf` directly and
restarting the service.

## 🔍 Checking the time sources {#checking-the-time-sources}

Check that the service is running, and that it actually has something to synchronize with:

```bash
systemctl status chronyd

> chronyc sources

Reviewer 2: Maybe you can mention the default/expected result  is:

```
systemctl status chronyd

   Active: active (running) since Tue 2026-08-11 21:26:08 GMT; 9h ago

chronyc sources
210 Number of sources = 4
MS Name/IP address         Stratum Poll Reach LastRx Last sample               
===============================================================================
^+ (...)                      1  10   377   276   +320us[ +320us] +/-   17ms
(...)
```

Author: Added, with the output taken off an 8.3 host here rather than composed. Running it turned up something I'd missed: the legend listed `*`, `+` and `?`, but a normal four-source listing is mostly `-`, which wasn't explained anywhere. So `-` is in the list now, along with the `^` source-type character.

---
sidebar_position: 8
---

# Time synchronization

XCP-ng keeps the dom0 clock synchronized using `chronyd`. A host whose date is wrong keeps
working normally until something needs to trust that date. When that happens, the resulting
failures rarely point to the clock as the cause. This page explains why the clock matters, how
to configure NTP, and how to check and correct it.

## ⏰ Why the clock must be correct {#why-the-clock-must-be-correct}

- **Access to the repositories.** Updates are fetched over HTTPS, and TLS validates the
  server's certificate against the host's own date. If the host's clock is set too far in the
  past, the TLS handshake fails because, from the host's perspective, the mirror's certificate
  is not yet valid. The error `yum` prints in that situation does not mention time.

> - **Pool membership.** A host joining a pool must have its clock synchronized with the pool

Reviewer 3: I think this would be worth expanding on. Same time when joining is necessary, but we also want it not to drift later, as this can cause subtle synchronization issues, and I'm not sure it's entirely covered by the next bullet point about certificates (which are used by hosts to authenticate to each other, notably).

Author: I expanded it. The bullet now covers staying synchronized as well as joining, and says why the second is the harder one to catch: a failed join tells you what is wrong; drift afterwards does not. It surfaces as hosts disagreeing about when things happened, each convinced by its own clock.

I kept it separate from the certificates bullet, since that is a different mechanism, and folded the old Pools section into this bullet as well, per your comment further down.

  master. See [Pool Requirements](../../installation/requirements#pool-requirements).
- **Certificates.** Certificates are only valid inside a validity window. A host whose date
  falls outside that window will reject, or be rejected by, connections that would otherwise
  succeed.
- **Logs and scheduled operations.** Correlating events across the hosts of a pool, and
  anything that runs on a schedule such as backups, depend on the hosts agreeing on what
  time it is.

## 🕰️ How XCP-ng keeps time {#how-xcp-ng-keeps-time}

The time daemon on dom0 is `chronyd`, configured through `/etc/chrony.conf`.

The default configuration enables `rtcsync`, allowing chronyd to keep the hardware clock
synchronized with the system clock automatically. There is no need to manually update the
hardware clock after correcting the system time, and the correction persists across reboots.

## 🔧 Configuring the time sources {#configuring-the-time-sources}

### During installation

> At step 11, the installer prompts you to enter the time, select a time zone and either

Reviewer 3: A reference to a numbered step in another page won't age well.

Author: I dropped the number. It now reads that the installer prompts for the time, the time zone, and either NTP servers or a manual time, and links to the install page without pointing to a position in it.

configure NTP servers or set the time manually. Always configure at least one NTP server. See
[Install XCP-ng](../../installation/install-xcp-ng).

:::warning
If you set the time manually at this step, the host can end up with `chronyd` running but no

> time source configured at all. It will keep whatever date it was installed with. See

Reviewer 3: "It will keep whatever date it was installed with." might need rewording, as it may let some think that the date will not increment every day.

Reviewer: Also, "it" seems ambiguous to me, in that sentence.

Author: I reworded it, and "it" is gone. It now says the clock still ticks, but nothing ever checks it, so it drifts from the date you typed, and if that date was wrong, it stays wrong.

Testing another of your comments turned up a reason to say more here: this is the one case that does not fix itself at boot. A host with sources gets stepped back into line within seconds of chronyd starting, so no sources at all is what makes this state persist. The warning says that now.

[Checking the time sources](#checking-the-time-sources) below.
:::

### During an automated installation

The installation answer file accepts one or more NTP servers through the `<ntp-server>`
element. See [Answer file](../../appendix/answerfile).

### On an installed host

> You can configure NTP from `xsconsole`, or by editing `/etc/chrony.conf` directly and

Reviewer 3: ⚠️ We discourage manually modifying configuration files that are maintained by XAPI. 

This should be of last resort if no other way works, such as xsconsole, `xe` or Xen Orchestra (does it support setting up NTP?).

Author: I changed it. `xsconsole` is the documented route now, and editing the file by hand is marked as a last resort in both places it appeared.

Regarding your question, I could be wrong, but I think XO does not configure NTP on hosts. `xoa network ntp` sets the time servers of the XOA appliance, which is a Debian VM keeping its own clock through `systemd-timesyncd`, and the NTP field in the deployment form does the same for the appliance being deployed:

```
@xen-orchestra/lite/src/pages/xoa-deploy.vue:447
  xapi.call('VM.add_to_xenstore_data', [vmRef.value, 'vm-data/ntp', ntp.value])
```

I grepped a clone of the tree rather than trusting a code search. Every hit outside `docs/` is that one locale string repeated per language plus the form using it, and `chrony` appears nowhere. 🤷 
The page now states the distinction, because appliance NTP and host NTP are easy to conflate.

systemctl status chronyd
chronyc sources
```

Both matter, and the second is the one that gets skipped. `chronyd` can be running and
enabled while having **no time sources configured at all**:

```
# chronyc sources
210 Number of sources = 0
```

Then, the service behaves exactly as configured, and does nothing. Every other check an
operator would normally run looks healthy, which is what makes this one worth running early
rather than late.

On a host that is synchronizing normally, the same two commands look like this:

```
# systemctl status chronyd
● chronyd.service - NTP client/server
   Loaded: loaded (/usr/lib/systemd/system/chronyd.service; enabled; vendor preset: enabled)
   Active: active (running) since Wed 2026-08-12 08:38:14 CEST; 46min ago

# chronyc sources
210 Number of sources = 4
MS Name/IP address         Stratum Poll Reach LastRx Last sample
===============================================================================
^* vps1.websters-computers.>     2   7   377    59   -361us[ -444us] +/- 8522us
^+ meshflow.net                  3   9   377   118   +203us[ +122us] +/-   12ms
^+ mail.rapidooo.fr              2   9   377   113  -1029us[-1110us] +/- 9934us

> ^- 88-185-213-3.subs.proxad>     3   8   377   506   +688us[ +491us] +/-   49ms

Reviewer 3: Suggestion: I'd replace the server names with placeholders, so that no one starts wondering why their results differ from the example.

Author: Done: `ntp1.example.net` through `ntp4.example.net`.

One tension worth naming: @rzr asked earlier for the real expected output, and this trades against that a little.
I kept every column value real and changed only the names, so the shape someone compares against is still a genuine listing. 🤷 

I'm happy to put the real hostnames back if you would rather.

```

In the `MS` column, the first character is the source type, `^` for a server, and the second
is the state of that source:

- `*` marks the one currently being used.
- `+` marks another acceptable source, combined with it.
- `-` marks a source chrony is measuring but has excluded from the combination.
- `?` means the source is unreachable.

At least one source should reach the `*` state. Seeing `-` on some of the others is normal
and does not need correcting.

To see the offset chrony believes it has, and whether it has settled:

```bash
chronyc tracking
```

## 🛠️ Correcting a wrong clock {#correcting-a-wrong-clock}

Start with `chronyc sources`, because the answer decides which of the two paths below
applies.

### If sources are listed

Correct the clock and check the result:

```bash
chronyc makestep
date
```

`chronyc makestep` is what does the work here. By default, chrony corrects time offsets by

> gradually slewing the system clock, which never converges for an offset of months or years.

Reviewer 3: I'm not sure about this. Either the service is healthy and this is unneeded, or it's the service that needs to be fixed, isn't it? Manually setting the clock once doesn't look like a long term solution and may just postpone the actual resolution by weeks or months.

I think it could be mentioned as an emergency fix, but not as a solution. And to add steps to go towards an actual solution (re-run xsconsole? check chronyd logs?).

It can also be useful as a one time check: "is chronyc able to fetch the date?", but I think it should be included in a higher level resolution process.

Unless I'm mistaken, it's XAPI that manages chronyc (worth verifying though), so the resolution could be at several layers:
* configuration layer (no sources configured, user mistake initially)
* XAPI layer (xsconsole being a XAPI client)
* chronyd service
* network, firewall (I don't see the later mentioned, but I would be surprised if that isn't a common cause).

Btw, I'm pretty sure Vates' support team would have interesting input on this.

Author: Testing your other comment showed that a host with reachable sources does not usually have a wrong clock at all: stock config has `makestep 1.0 3`, and a 45-day offset was corrected within ten seconds of chronyd starting. So a clock that is still wrong is evidence that something upstream is broken, and stepping it by hand mostly hides that. 🤔 

Restructured along the layers you listed, working outward:

1. Any sources at all, `chronyc sources`, with `xsconsole` to add them.
2. Sources reachable. Everything stuck at `?` is configured but unanswered, so UDP 123 and the firewall get named here, along with the no-route-to-the-pool case.
3. Daemon health: `systemctl status chronyd` and `journalctl -u chronyd`.
4. Configuration changed underneath you, which `rpm -V chrony` will show.

Stepping the clock is its own short section now, described as a stopgap that buys time for the diagnosis rather than ending it, including the detail that it reports success while moving nothing when there are no sources.

One correction on XAPI managing chronyc, since you flagged it as worth verifying: `/etc/chrony.conf` is owned by the `chrony` package, and nothing under `/opt/xensource` writes it.
xsconsole is an XAPI client that edits the file and restarts the daemon, which comes to the same thing for a user, but it does mean `rpm -V chrony` reports its edits as local modifications.

On the support team, I have not asked them. Their ordering of these layers would be better than mine, since they see which one actually turns out to be the cause.

### If `Number of sources = 0`

Do not start with `makestep`. It will report success and move nothing, because chrony has no
measured offset to step to.

1. Add time sources to `/etc/chrony.conf`:

   ```
   server 0.pool.ntp.org iburst
   server 1.pool.ntp.org iburst
   server 2.pool.ntp.org iburst

>    server 3.pool.ntp.org iburst

Reviewer 3: again, editing the configuration file directly should be last resort.

Author: Same change as above, applied here too. Both places now put `xsconsole` first and mark editing the file by hand as a last resort.

   ```

2. Restart the service and confirm that a source has become reachable, as described in
   [Checking the time sources](#checking-the-time-sources):

   ```bash
   systemctl restart chronyd
   chronyc sources
   ```

3. Once a source is reachable, correct the clock:

   ```bash
   chronyc makestep
   date
   ```

:::note
Correcting the date does not invalidate the host's own certificate. XAPI issues it with a
ten-year validity, so a host installed with a wrong date still holds a certificate that

> covers the corrected date. `xe host-refresh-server-certificate` is not needed for this.

Reviewer 3: "a host installed with a wrong date still holds a certificate that covers the corrected date"

Can we affirm that? It's likely, not guaranteed.

Also, this note brings up a topic that many users wouldn't have thought of initially, only to dismiss it directly. I'd rather 1. tell them that they're probably fine but 2. also tell them how to verify and 3. how to fix it if really necessary (if we're able to provide the appropriate steps, validated).

Author: I read one off an 8.3 host:

```
notBefore=Jun  2 15:17:54 2026 GMT
notAfter=May 30 15:17:54 2036 GMT
```

Ten years wide, anchored on the clock at the moment it was generated. A host that installs itself believing the year is 2010 gets a certificate valid from 2010 to 2020, so correcting the clock to now leaves it expired. Set the clock far forward at install, and you get the mirror image: not valid yet once you correct backwards. My note leaned on the width, and the anchor is what actually decides it.

Rewritten along your three lines. Most corrections are fine, and it says so. To check:

```bash
openssl x509 -in /etc/xensource/xapi-ssl.pem -noout -dates
date -u
```

To fix, if the date falls outside that window, `xe host-refresh-server-certificate host=<host>`, with `xe host-emergency-reset-server-certificate` for a host already being rejected and therefore not reachable the usual way. Both were checked against `xe help` on the host.

Author: Correction to my own reply above, and a better answer than the one I gave.

I reproduced it on a lab host, and the result contradicted part of what I wrote. On an 8.3 host, chronyd stopped and the clock moved to 2016:

```
xe host-refresh-server-certificate host=<uuid>

/etc/xensource/xapi-pool-tls.pem
  notBefore=Aug 17 10:00:00 2016 GMT
  notAfter=Aug 15 10:00:00 2026 GMT
```

Ten years, anchored on the fake clock. Restoring the real date left that certificate expired by two days.

What I had wrong: `xapi-ssl.pem` never moved during any of this. `host-refresh-server-certificate` renews the internal pool certificate, and `xe help` says so plainly once read, "Refresh internal server certificate of host", against "Deletes the current TLS server certificate in the host" for `host-reset-server-certificate`. So my note told the reader to check the dates on one certificate and then run the command that renews the other. It would have reported success and changed nothing, which is the worst shape for a troubleshooting instruction.

Fixed in `c848a9b`. The note now lists both certificates with the command that regenerates each, tells the reader to check both windows, and warns against exactly the substitution I had just made.

The host is back to a 2026 to 2036 pool certificate and healthy. I will open an issue on `xcp-ng/xcp` for the underlying behaviour, since a certificate anchored to a wrong install-time clock is a product question rather than a documentation one, and link it here.

:::

:::tip
If `date` is wrong again after every power cycle, the motherboard's RTC battery is probably
dead and should be replaced. `rtcsync` can only keep the hardware clock in step while the

> host is running; it cannot help a clock that loses its value when the power goes.

Reviewer 3: What consequences does it have? Doesn't chronyd fix the date shortly after initial startup? Does it cause problems at startup that do not self-heal?

Author: Yes, chronyd fixes it quickly, as long as it has a source.

The stock `/etc/chrony.conf` carries `makestep 1.0 3`, so for the first three updates after the service starts, it jumps the clock instead of easing it into place. I stopped chronyd, put the clock 45 days into the past, and started it again:

```
skew    2026-07-03T12:17:41Z   chronyd started
t+05s   2026-07-03  still wrong
t+10s   2026-08-17T12:17:46Z   corrected
```

Under ten seconds for 45 days. The page claimed chrony only slews and so never converges for offsets of months or years, which is true once those three updates are spent and wrong at startup. Fixed, and it changes the shape of the page: a host with sources heals itself, so the manual correction is for the daemon that has been up a while, or the host that never had a source at all.

What does not self-heal is `Number of sources = 0`, since `makestep` needs something to measure against. That is now said where the manual-time-at-install warning is.

Reviewer 3: > If `date` is wrong again after every power cycle

This part implies, to me, that we you get a wrong date again after every power cycle DESPITE ntp, and I don't see how that is possible thanks to makestep, unless there's a problem with the ntp configuration which must then be fixed and the battery then doesn't matter. Or does it? 

It's still unclear to me what the problem is if you have a dead battery, outside the fact that the clock needs to be fixed at each boot (implying a part of the boot process with wrong log times and maybe worse consequences that I can't foresee).

:::

## 🌐 Networks without internet access {#networks-without-internet-access}

Hosts without Internet access cannot reach the public NTP pool. The result is the same as
having no sources at all: `chronyc sources` lists servers that never leave the `?` state.

On such networks, point `/etc/chrony.conf` to a time source that hosts can actually reach,
such as an appliance on the same network, or a local server that is itself synchronized and
acts as the reference time for the network.

> ## 🎱 Pools {#pools}

Reviewer 3: I'm not sure this section brings much value besides what's already described in "Why the clock must be correct", at least in its current state.

Author: Agreed, it was saying the same thing twice. Folded into the Pool membership bullet in "Why the clock must be correct" and dropped the section.

The merge also gave that bullet somewhere to make your point from the other comment, that we care about drift after joining and not only about the clock at join time. Nothing referenced the `#pools` anchor, so no links break.

## 🕰️ How XCP-ng keeps time {#how-xcp-ng-keeps-time}

The time daemon on dom0 is `chronyd`, configured through `/etc/chrony.conf`.

The default configuration enables `rtcsync`, allowing chronyd to keep the hardware clock
synchronized with the system clock automatically. There is no need to manually update the
hardware clock after correcting the system time, and the correction persists across reboots.

It also sets `makestep 1.0 3`, which lets chronyd jump the clock rather than ease it into
place, for the first three updates after the service starts. A host that boots with a badly
wrong date therefore fixes itself within seconds, as long as it can reach a time source.
Measured on 8.3: a clock put 45 days in the past was corrected between five and ten seconds
after `chronyd` started.

After those three updates chronyd only slews, which for an offset of months or years never
converges. That is the case the manual correction below exists for, along with the host that
had no time source to begin with.

## 🔧 Configuring the time sources {#configuring-the-time-sources}

### During installation

The installer prompts you to enter the time, select a time zone, and either configure NTP

> servers or set the time manually. Always configure at least one NTP server. See

Reviewer: This might be a nitpick, but _"See Install XCP-ng"_  feels disconnected from the previous sentence. Linking the sentence closer to the previous one would help understand its meaning. For example, we can write one of these:

- _"Always configure at least one NTP server (see Install XCP-ng)"_ (all in the same sentence), or:
- _"Always configure at least one NTP server. See Install XCP-ng for more information."_ (keep the two-sentence structure, but add a phrase in the second sentence that semantically links it to the first)

My comment also applies to the other occurrences of "See X" on the same page.

Author: Applied to all five, in `1dbc53c`. I used your second shape where the sentence had room for it and the parenthetical where it did not:

- Pool Requirements: what else a host must meet before it can join
- Install XCP-ng: where this falls in the installation sequence
- Checking the time sources: how to recognise that state on a running host
- Answer file: its syntax, as a parenthetical
- Networks without internet access: folded into the clause about no route to the pool, so it is no longer a separate sentence at all

Two other uses of "see" on the page are ordinary verbs, "to see the offset chrony believes it has" and "the TLS certificate clients see", so I left those.

Writing the connecting phrases was a decent check on the links themselves: each one made me say what the reader would find on the other side, and the Networks one turned out not to need a cross-reference at all once the sentence said where the problem was covered.

---
sidebar_position: 8
---

# Time synchronization

XCP-ng keeps the dom0 clock synchronized using `chronyd`. A host whose date is wrong keeps
working normally until something needs to trust that date. When that happens, the resulting
failures rarely point to the clock as the cause. This page explains why the clock matters, how
to configure NTP, and how to check and correct it.

## Why the clock must be correct {#why-the-clock-must-be-correct}

- **Access to the repositories.** Updates are fetched over HTTPS, and TLS validates the
  server's certificate against the host's own date. If the host's clock is set too far in the
  past, the TLS handshake fails because, from the host's perspective, the mirror's certificate
  is not yet valid. The error `yum` prints in that situation does not mention time.

> - **Pool membership.** A host joining a pool must have its clock synchronized with the pool
>   master, and must stay synchronized afterwards. A failed join tells you what is wrong. Drift
>   that sets in later does not: it surfaces as hosts disagreeing about when things happened,
>   each of them convinced by its own clock. Keep every host in a pool on the same time sources.
>   See [Pool Requirements](../../installation/requirements#pool-requirements) for the other
>   conditions a host has to meet before it can join.

Reviewer 3: Suggested rewording. Straight out of my fingers, so feel free to improve the language correctness.

```suggestion
- **Hosts synchronization inside a pool.** Desynchronized time references between hosts
  in a pool can cause a variety of subtle issues for operations that require the hosts to agree
  on the current time. As such, it is necessary for hosts to synchronize time from a common
  source.  Disagreeing clocks can also prevent adding a new host to an existing pool. The correct
  way to fix it is to ensure synchronization works, rather than a one time fix which would let
  them drift apart in the future.
```

- **Certificates.** Certificates are only valid inside a validity window. A host whose date
  falls outside that window will reject, or be rejected by, connections that would otherwise
  succeed.
- **Logs and scheduled operations.** Correlating events across the hosts of a pool, and
  anything that runs on a schedule such as backups, depend on the hosts agreeing on what
  time it is.

> 

Reviewer 3: Here let's add a word about Xen Orchestra, which, when used, must also have a synchronized clock?

## How XCP-ng keeps time {#how-xcp-ng-keeps-time}

The time daemon on dom0 is `chronyd`, configured through `/etc/chrony.conf`.

The default configuration enables `rtcsync`, allowing chronyd to keep the hardware clock
synchronized with the system clock automatically. There is no need to manually update the
hardware clock after correcting the system time, and the correction persists across reboots.

It also sets `makestep 1.0 3`, which lets chronyd jump the clock rather than ease it into
place, for the first three updates after the service starts. A host that boots with a badly
wrong date therefore fixes itself within seconds, as long as it can reach a time source.
Measured on 8.3: a clock put 45 days in the past was corrected between five and ten seconds

> after `chronyd` started.

Reviewer 3: Let's remove this log-like precision.

> After those three updates chronyd only slews, which for an offset of months or years never

Reviewer 3: I didn't know verb "to slew", and think that "gradually adjusts" who be easier to understand by non-native english speakers like me.

converges. That is the case the manual correction below exists for, along with the host that

> had no time source to begin with.

Reviewer 3: I'm not so sure about this claim. As I understand it, a reboot fixes the clock if ntp is correctly configured, so:
- either the hosts haven't been rebooted in a very long time and they're likely behind on updates, and should probably be updated+rebooted.
- if rebooting doesn't fix the problem, it's probably elsewhere and not caused by the fact that chronyd can only gradually adjust time after first three times since last reboot, but by another, bigger cause (heavily drifting hardware clock, misconfigured ntp or failure to join it due to firewall...).

## Configuring the time sources {#configuring-the-time-sources}

### During installation

The installer prompts you to enter the time, select a time zone, and either configure NTP
servers or set the time manually. Always configure at least one NTP server (see
[Install XCP-ng](../../installation/install-xcp-ng) for where this falls in the installation
sequence).

:::warning
If you set the time manually at this step, the host can end up with `chronyd` running but no
time source configured at all. The clock still ticks, but nothing ever checks it: it drifts
from the date you typed, and if that date was wrong, it stays wrong. This is also the one

> case that does not fix itself at boot, since `makestep` needs a source to measure against.

Reviewer 3: "This is also the one case that does not fix itself at boot, since `makestep` needs a source to measure against."

=> this could be removed, if the sentence just above would say that nothing adjusts it (instead of checks it).

- `?` means the source is unreachable.

At least one source should reach the `*` state. Seeing `-` on some of the others is normal
and does not need correcting.

To see the offset chrony believes it has, and whether it has settled:

```bash
chronyc tracking
```

## Correcting a wrong clock {#correcting-a-wrong-clock}

A host with working time sources does not usually have a wrong clock, because chronyd steps
it within seconds of starting. So a clock that is still wrong is telling you that something
upstream of it is broken. Find that first. Stepping the clock by hand gets you working again
today, and if you stop there the date is wrong again in a few weeks.

Work outward, from the configuration to the network:

1. **Are there any sources?** `chronyc sources`. `Number of sources = 0` means none were
   ever configured, which is the manual-time-at-install case. Add them with `xsconsole`.
2. **Are they reachable?** Sources stuck at `?` are configured but unanswered. NTP goes out
   over UDP 123, so a firewall between the host and its servers produces exactly this, as
   does a network with no route to the public pool, which

>    [Networks without internet access](#networks-without-internet-access) below covers.

Reviewer 3: It could also be that the ntp server they have chosen doesn't exist anymore

3. **Is the daemon healthy?** `systemctl status chronyd`, and `journalctl -u chronyd` for
   what it made of its configuration at startup.
4. **Was the configuration changed underneath you?** `rpm -V chrony` reports whether

>    `/etc/chrony.conf` still matches the package. Hand edits and `xsconsole` both show up here.

Reviewer 3: I have doubts about this one. Can you verify?

1. If xsconsole changes chrony.conf, then of course it will differ from the base file provided by the package
2. more generally, rpm -V checks differences between a file as provided by the RPM package and what's on the system. For a configuration file, it's natural that there's a difference. It's there for this purpose. I don't know what rpm -V says here but I wouldn't be surprised if either it would always show it different (with a flag stating that it's configuration), or always ignore it. It also depends whether the file is declared %config or %config(noreplace) in the RPM spec file. Here, we have `%config(noreplace) %{_sysconfdir}/chrony.conf` which means that the RPM must not overwrite the file with future updates if the sysadmin (XAPI/xsconsole being considered an extension of the sysadmin for this purpose) has made any change to the file. 

So, the output of rpm -V tells you nothing here, except maybe in some cases "it was never configured", but you would already have found out with earlier checks.

What you probably really want is to determine if something has modified the configuration outside the supported way (installer, xsconsole, xe/XAPI), and I'm not sure there's an answer to this one, but I haven't digged into the details.

Once sources are reachable, restarting the daemon corrects the clock on its own, because the

> `makestep` allowance applies again from a fresh start:

Reviewer 3: To understand this, one must have read #how-xcp-ng-keeps-time

```bash
systemctl restart chronyd
chronyc sources
date
```

### Stepping the clock immediately

If you need the correct date right now and the daemon has been running for a while, its

> `makestep` allowance is already spent and it will only slew:

Reviewer 3: Same comment about "slew" that earlier.

expired. A clock set far into the future produces the mirror image: certificates not valid
yet once the date is corrected backwards.

There are two, and they are refreshed by different commands:

| Certificate | What it is | Regenerate with |
|---|---|---|
| `/etc/xensource/xapi-ssl.pem` | The TLS certificate clients see, including Xen Orchestra | `xe host-reset-server-certificate` |
| `/etc/xensource/xapi-pool-tls.pem` | The internal certificate hosts use between themselves | `xe host-refresh-server-certificate host=<host>` |

Check both windows against the corrected date:

```bash
openssl x509 -in /etc/xensource/xapi-ssl.pem -noout -dates
openssl x509 -in /etc/xensource/xapi-pool-tls.pem -noout -dates
date -u
```

If the date falls outside `notBefore` to `notAfter`, regenerate the one concerned using the
table above. Do not reach for `host-refresh-server-certificate` to fix `xapi-ssl.pem`: it
refreshes the internal certificate, reports success, and leaves the TLS one untouched.

On a host whose TLS certificate is already rejected, and which therefore cannot be reached the
usual way, `xe host-emergency-reset-server-certificate` runs locally on the host itself.

> :::

Reviewer 3: I think this should just be a short note explaining how to verify the certificate validity periods. How to renew certificates is a topic that doesn't belong here. There's an old draft on this whole topic, which stalled: https://github.com/xcp-ng/xcp-ng-org/pull/216

The certificate topic is a big one in itself, we should cover it independently.

:::tip
If `date` is wrong again after every power cycle, the motherboard's RTC battery is probably
dead and should be replaced. `rtcsync` can only keep the hardware clock in step while the
host is running; it cannot help a clock that loses its value when the power goes.
:::

## Networks without internet access {#networks-without-internet-access}

Hosts without Internet access cannot reach the public NTP pool. The result is the same as
having no sources at all: `chronyc sources` lists servers that never leave the `?` state.

On such networks, point the hosts at a time source they can actually reach, using `xsconsole`
as above: an appliance on the same network, or a local server that is itself synchronized and

> acts as the reference time for the network.

Reviewer 3: using xsconsole is one way to do it, for an already installed host. But here we're not in the troubleshooting part of the document anymore, so the best course of action is to set it right at installation time.
